### PR TITLE
Final Merge of Push Contracts 2.5 Dev New structure and tests

### PR DESCRIPTION
**Pull request by zaryab2000**
This PR is a mega merge of the new structure of the push contracts along with other files such as tests, interfaces, libraries etc.

Some of the key updates in this PR are as follows:

- **Feature:** Auto-subscription of delegates when added to a channel
- Test Migration of Hardhat to Foundry
- Fuzz tests for Core and Staking
- Interfaces for Core, Comm, Staking contracts
- Inclusion of Custom Errors for all contracts
- Changes in Deployment scripts w.r.t OZ-5
- Inclusion of libraries and Separate files for Custom Data Type handling/management
- Finally, moved all updated foundry test to **test** folder while the old hardhat tests can be found under **test_legacy** folder

### CODE REVIEW

File: test/PushComm/unit_tests/SendingNotifications/SendNotifs.t.sol [hunk 1]
```diff
@@ -0,0 +1,71 @@
+pragma solidity ^0.8.0;
+
+import { BaseTest } from "../../../BaseTest.t.sol";
+import { CoreTypes } from "../../../../contracts/libraries/DataTypes.sol";
+
+// import "forge-std/console.sol";
+contract SendNotifs_Test is BaseTest {
+    bytes constant _testChannelIdentity = bytes("test-channel-hello-world");
```
Comment (0xNilesh) [inline test/PushComm/unit_tests/SendingNotifications/SendNotifs.t.sol:8]: this variable is used in all 3 files of comm tests, and in BasePushCoreTest.t.sol
If it's getting used in whole test codebase, then it's better to put it in BaseTest itself

File: test/PushComm/unit_tests/SendingNotifications/SendNotifs.t.sol [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    }
+
+    function test_WhenAUserIsSendingNotifToOtherAddressInsteadOfThemselves() external {
+        changePrank(actor.charlie_channel_owner);
+        bool res = commProxy.sendNotification(actor.bob_channel_owner, actor.alice_channel_owner, _testChannelIdentity);
+        assertEq(res, false);
+    }
+
+    function test_WhenRecipientIsSendingNOTIFOnlyToThemselves() external {
+        //User can no longer send notifs to themselves
```
Comment (0xNilesh) [inline test/PushComm/unit_tests/SendingNotifications/SendNotifs.t.sol:23]: this test is missed

File: test/PushComm/unit_tests/SendingNotifications/SendNotifsViaSig.t.sol [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+    struct SendNotif {
+        address _channel;
+        address _recipient;
+        bytes _identity;
+        uint256 nonce;
+        uint256 expiry;
+    }
+
+    // computes the hash of a sendNotif
+    function getStructHash(SendNotif memory _sendNotif) internal pure returns (bytes32) {
```
Comment (0xNilesh) [inline test/PushComm/unit_tests/SendingNotifications/SendNotifsViaSig.t.sol:193]: There could be a BasePushCommTest file that could have these extra helper fn and variables

File: test/PushComm/unit_tests/SendingNotifications/SendNotifsViaSig.t.sol [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+            v,
+            r,
+            s
+        );
+
+        assertEq(false, res);
+    }
+
+    //HELPER FUNCTIONS
+    struct SendNotif {
```
Comment (0xNilesh) [inline test/PushComm/unit_tests/SendingNotifications/SendNotifsViaSig.t.sol:184]: this one in BasePushCommTest as well

File: test/PushComm/unit_tests/SendingNotifications/SendNotifsViaSig.t.sol [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+                _sendNotif._channel,
+                _sendNotif._recipient,
+                keccak256(_sendNotif._identity),
+                _sendNotif.nonce,
+                _sendNotif.expiry
+            )
+        );
+    }
+
+    function getDomainSeparator() public returns (bytes32) {
```
Comment (0xNilesh) [inline test/PushComm/unit_tests/SendingNotifications/SendNotifsViaSig.t.sol:206]: this one in BasePushCommTest as well

File: test/PushComm/unit_tests/SubscribeBySig/SubscribeBySig.t.sol [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+import { BaseTest } from "../../../BaseTest.t.sol";
+import { CoreTypes } from "../../../../contracts/libraries/DataTypes.sol";
+import { SignatureVerifier } from "contracts/mocks/MockERC1271.sol";
+
+import "forge-std/console.sol";
+
+contract SubscribeBySig_Test is BaseTest {
+    bytes constant _testChannelIdentity = bytes("test-channel-hello-world");
+
+    event Subscribe(address indexed channel, address indexed user);
```
Comment (0xNilesh) [inline test/PushComm/unit_tests/SubscribeBySig/SubscribeBySig.t.sol:12]: these events should've been added in test/utils/Events.sol file under CommEvents

File: test/PushComm/unit_tests/SubscribeBySig/SubscribeBySig.t.sol [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+    event Unsubscribe(address indexed channel, address indexed user);
```
Comment (0xNilesh) [inline test/PushComm/unit_tests/SubscribeBySig/SubscribeBySig.t.sol:13]: this too

File: test/PushComm/unit_tests/SubscribeBySig/SubscribeBySig.t.sol [hunk 8]
```diff
@@ -0,0 +1,294 @@
+pragma solidity ^0.8.0;
+
+import { BaseTest } from "../../../BaseTest.t.sol";
+import { CoreTypes } from "../../../../contracts/libraries/DataTypes.sol";
+import { SignatureVerifier } from "contracts/mocks/MockERC1271.sol";
+
+import "forge-std/console.sol";
```
Comment (0xNilesh) [inline test/PushComm/unit_tests/SubscribeBySig/SubscribeBySig.t.sol:7]: unused import

File: test/PushComm/unit_tests/SubscribeBySig/SubscribeBySig.t.sol [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+
+        changePrank(actor.bob_channel_owner);
+        coreProxy.createChannelWithPUSH(CoreTypes.ChannelType.InterestBearingOpen, _testChannelIdentity, 50e18, 0);
+    }
+
+    modifier whenUserSubscribesWith712Sig() {
+        _;
+    }
+
+    function testFail_WhenMaliciousUserReplaysA712Signature() public whenUserSubscribesWith712Sig {
```
Comment (0xNilesh) [inline test/PushComm/unit_tests/SubscribeBySig/SubscribeBySig.t.sol:31]: is this issue of expectRevert remaining?
and if not, then I think we should remove testFail and use expectRevert, as it's the case in all tests

File: test/PushComm/unit_tests/SubscribeBySig/SubscribeBySig.t.sol [hunk 10]
```diff
@@ -0,0 +1,294 @@
...
+        uint256 oldNonce = commProxy.nonces(actor.alice_channel_owner) - 1;
+        changePrank(actor.charlie_channel_owner);
+        // Revert fails for unknown reason - Needs to be fixed
+        //vm.expectRevert(bytes("PushCommV2::subscribeBySig: Invalid nonce"));
+        commProxy.subscribeBySig(
+            actor.bob_channel_owner, actor.alice_channel_owner, oldNonce, block.timestamp + 1000, v, r, s
+        );
+    }
+
+    function test_WhenAnExpired712SignatureIsPassed() public whenUserSubscribesWith712Sig {
```
Comment (0xNilesh) [inline test/PushComm/unit_tests/SubscribeBySig/SubscribeBySig.t.sol:61]: as this is a test that has expectRevert, this should've been named as test_Revertwhen_
It's good to maintain uniformity across all tests

It's the case with few tests in this file, all should be changed accordingly

File: test/PushComm/unit_tests/SubscribeBySig/SubscribeBySig.t.sol [hunk 11]
```diff
@@ -0,0 +1,294 @@
...
+    //Helper Functions
+    struct SubscribeUnsubscribe {
+        address _channel;
+        address _subscriber;
+        uint256 nonce;
+        uint256 expiry;
+    }
+
+    // computes the hash of a sendNotif
+    function getStructHash(
```
Comment (0xNilesh) [inline test/PushComm/unit_tests/SubscribeBySig/SubscribeBySig.t.sol:272]: these can be put in BasePushCommTest file as well

**Review by 0xNilesh — CHANGES_REQUESTED** [review]

**Review by 0xNilesh — CHANGES_REQUESTED** [review]
When I run `forge test`,
It's showing these errors

image

File: contracts/interfaces/IPushCoreStaking.sol [hunk 12]
```diff
@@ -1,7 +1,9 @@
 pragma solidity ^0.8.20;
 
-interface IPushCore {
-    function handleChatRequestData(address requestSender, address requestReceiver, uint256 amount) external;
-    function PROTOCOL_POOL_FEES() external view returns (uint256);
+import { CoreTypes } from "../libraries/DataTypes.sol";
```
Comment (Zartaj0) [inline contracts/interfaces/IPushCoreStaking.sol:3]: Unused import

**Review by Zartaj0 — APPROVED** [review]

**Comment by zaryab2000** [comment]
Merged this branch finally to pushContracts_2.5_dev branch.

**Pull request merged**